pkg/prom/exporter.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prom
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	retrieveDurationDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("onos", "exporter", "retriever_duration_seconds"),
+		"onos_exporter: retrieve duration.",
+		[]string{"retriever"},
+		nil,
+	)
+	retrieveSuccessDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("onos", "exporter", "retriever_success"),
+		"onos_exporter: retriever succeeded.",
+		[]string{"retriever"},
+		nil,
+	)
+)
+
+// Collector Interface that defines the expected method of a Collector
+// Retrieve method must send prometheus metrics to the ch channel
+type Collector interface {
+	Retrieve(ch chan<- prometheus.Metric) error
+}
+
+// Exporter Defines a exporter struct in a address/path and with a set of collectors
+type Exporter struct {
+	path       string
+	address    string
+	Collectors map[string]Collector
+}
+
+// RegisterCollector Registers Collectors to a Prometheus exporter
+func (e Exporter) RegisterCollector(collectorName string, collector Collector) {
+	e.Collectors[collectorName] = collector
```

Comment:
this write is not thread safe. Perhaps it would be better to move Collectors out of the exporter and implement it as interface and then use mutex for Write and Read and then inject it into the Exporter. example: https://github.com/onosproject/onos-e2t/blob/master/pkg/ranfunctions/registry.go

Reply:
Indeed, I was only considering all the collectors are defined prior to the exporter Run.